File: crates/secd-ui/src/crypto.rs
//! Client AEAD and wraps. Same locked params as secd-core. JS cannot zeroize; we overwrite.
//! The DEK lives only in a Zeroizing signal for the tab's lifetime — never in
//! storage — so a reload drops it until the next sign-in.

use argon2::{Algorithm, Argon2, Params, Version};
use chacha20poly1305::aead::rand_core::RngCore;
use chacha20poly1305::aead::{Aead, KeyInit, OsRng, Payload};
use chacha20poly1305::{XChaCha20Poly1305, XNonce};
use serde_json::{json, Value};
use x25519_dalek::{PublicKey, StaticSecret};
use zeroize::Zeroize;

const NONCE_LEN: usize = 24;
const KEY_LEN: usize = 32;
const SALT_LEN: usize = 16;
const ARGON2_M_KIB: u32 = 19_456;
const ARGON2_T: u32 = 2;
const ARGON2_P: u32 = 1;

#[derive(Clone, Debug)]
pub struct Wrap {
    pub factor: Factor,
    pub cred_id: Option<String>,
    pub salt: Option<String>,
    pub blob: String,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Factor {
    Passkey,
    Password,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum CryptoError {
    Key,
    Aead,
    Truncated,
    Hex,
    Rng,
    Prf,
    Factor,
}

pub fn mint_dek() -> [u8; KEY_LEN] {
    let mut dek = [0u8; KEY_LEN];
    OsRng.fill_bytes(&mut dek);
    dek
}

pub fn zeroize_bytes(buf: &mut [u8]) {
    buf.zeroize();
}

pub fn seal(key: &[u8], name: &str, plaintext: &[u8]) -> Result<Vec<u8>, CryptoError> {
    let cipher = cipher(key)?;
    let mut nonce_bytes = [0u8; NONCE_LEN];
    OsRng
        .try_fill_bytes(&mut nonce_bytes)
        .map_err(|_| CryptoError::Rng)?;
    let nonce = XNonce::from_slice(&nonce_bytes);
    let ct = cipher
        .encrypt(
            nonce,
            Payload {
                msg: plaintext,
                aad: name.as_bytes(),
            },
        )
        .map_err(|_| CryptoError::Aead)?;
    let mut blob = Vec::with_capacity(NONCE_LEN + ct.len());
    blob.extend_from_slice(&nonce_bytes);
    blob.extend_from_slice(&ct);
    Ok(blob)
}

pub fn open(key: &[u8], name: &str, blob: &[u8]) -> Result<Vec<u8>, CryptoError> {
    if blob.len() < NONCE_LEN + 1 {
        return Err(CryptoError::Truncated);
    }
    let cipher = cipher(key)?;
    let nonce = XNonce::from_slice(&blob[..NONCE_LEN]);
    cipher
        .decrypt(
            nonce,
            Payload {
                msg: &blob[NONCE_LEN..],
                aad: name.as_bytes(),
            },
        )
        .map_err(|_| CryptoError::Aead)
}

fn cipher(key: &[u8]) -> Result<XChaCha20Poly1305, CryptoError> {
    if key.len() != KEY_LEN {
        return Err(CryptoError::Key);
    }
    XChaCha20Poly1305::new_from_slice(key).map_err(|_| CryptoError::Key)
}

pub fn wrap_password(dek: &[u8], password: &[u8]) -> Result<Wrap, CryptoError> {
    if dek.len() != KEY_LEN {
        return Err(CryptoError::Key);
    }
    let mut salt = [0u8; SALT_LEN];
    OsRng
        .try_fill_bytes(&mut salt)
        .map_err(|_| CryptoError::Rng)?;
    let mut kek = derive_password_kek(password, &salt)?;
    let blob = seal(&kek, "password", dek)?;
    kek.zeroize();
    let salt_hex = to_hex(&salt);
    salt.zeroize();
    Ok(Wrap {
        factor: Factor::Password,
        cred_id: None,
        salt: Some(salt_hex),
        blob: to_hex(&blob),
    })
}

pub fn unwrap_password(wrap: &Wrap, password: &[u8]) -> Result<Vec<u8>, CryptoError> {
    if wrap.factor != Factor::Password {
        return Err(CryptoError::Factor);
    }
    let salt_hex = wrap.salt.as_deref().ok_or(CryptoError::Factor)?;
    let mut salt = from_hex(salt_hex)?;
    if salt.len() != SALT_LEN {
        salt.zeroize();
        return Err(CryptoError::Hex);
    }
    let mut kek = derive_password_kek(password, &salt)?;
    salt.zeroize();
    let blob = from_hex(&wrap.blob)?;
    let dek = open(&kek, "password", &blob);
    kek.zeroize();
    dek
}

pub fn wrap_passkey(dek: &[u8], prf: &[u8], cred_id: &str) -> Result<Wrap, CryptoError> {
    if dek.len() != KEY_LEN {
        return Err(CryptoError::Key);
    }
    let mut kek = prf_kek(prf)?;
    let blob = seal(&kek, "passkey", dek)?;
    kek.zeroize();
    Ok(Wrap {
        factor: Factor::Passkey,
        cred_id: Some(cred_id.to_string()),
        salt: None,
        blob: to_hex(&blob),
    })
}

pub fn unwrap_passkey(wrap: &Wrap, prf: &[u8]) -> Result<Vec<u8>, CryptoError> {
    if wrap.factor != Factor::Passkey {
        return Err(CryptoError::Factor);
    }
    let mut kek = prf_kek(prf)?;
    let blob = from_hex(&wrap.blob)?;
    let dek = open(&kek, "passkey", &blob);
    kek.zeroize();
    dek
}

pub fn wrap_from_json(v: &Value) -> Option<Wrap> {
    let factor = match v.get("factor")?.as_str()? {
        "passkey" => Factor::Passkey,
        "password" => Factor::Password,
        _ => return None,
    };
    Some(Wrap {
        factor,
        cred_id: v
            .get("cred_id")
            .and_then(|x| x.as_str())
            .map(str::to_string),
        salt: v.get("salt").and_then(|x| x.as_str()).map(str::to_string),
        blob: v.get("blob")?.as_str()?.to_string(),
    })
}

pub fn wrap_to_json(w: &Wrap) -> Value {
    let mut m = serde_json::Map::new();
    let factor = match w.factor {
        Factor::Passkey => "passkey",
        Factor::Password => "password",
    };
    m.insert("factor".into(), json!(factor));
    if let Some(c) = &w.cred_id {
        m.insert("cred_id".into(), json!(c));
    }
    if let Some(s) = &w.salt {
        m.insert("salt".into(), json!(s));
    }
    m.insert("blob".into(), json!(w.blob));
    Value::Object(m)
}

pub fn wraps_from_json(v: &Value) -> Vec<Wrap> {
    v.get("wraps")
        .and_then(Value::as_array)
        .map(|arr| arr.iter().filter_map(wrap_from_json).collect())
        .unwrap_or_default()
}

pub fn unwrap_any(wraps: &[Wrap], password: Option<&[u8]>, prf: Option<&[u8]>) -> Option<Vec<u8>> {
    if let Some(prf) = prf {
        for w in wraps {
            if w.factor == Factor::Passkey {
                if let Ok(dek) = unwrap_passkey(w, prf) {
                    return Some(dek);
                }
            }
        }
    }
    if let Some(password) = password {
        for w in wraps {
            if w.factor == Factor::Password {
                if let Ok(dek) = unwrap_password(w, password) {
                    return Some(dek);
                }
            }
        }
    }
    None
}

pub fn seal_dek_to_eph(dek: &[u8], their_pub: &[u8]) -> Result<Value, CryptoError> {
    if dek.len() != KEY_LEN || their_pub.len() != KEY_LEN {
        return Err(CryptoError::Key);
    }
    let mut their = [0u8; 32];
    their.copy_from_slice(their_pub);
    let secret = StaticSecret::random_from_rng(OsRng);
    let public = PublicKey::from(&secret);
    let shared = secret.diffie_hellman(&PublicKey::from(their));
    let blob = seal(shared.as_bytes(), "dek", dek)?;
    Ok(json!({
        "alg": "x25519-xchacha20poly1305",
        "eph_pub": to_hex(public.as_bytes()),
        "blob": to_hex(&blob),
    }))
}

fn prf_kek(prf: &[u8]) -> Result<[u8; KEY_LEN], CryptoError> {
    if prf.len() < KEY_LEN {
        return Err(CryptoError::Prf);
    }
    let mut kek = [0u8; KEY_LEN];
    kek.copy_from_slice(&prf[..KEY_LEN]);
    Ok(kek)
}

fn derive_password_kek(password: &[u8], salt: &[u8]) -> Result<[u8; KEY_LEN], CryptoError> {
    let params = Params::new(ARGON2_M_KIB, ARGON2_T, ARGON2_P, Some(KEY_LEN))
        .expect("invariant: locked argon2id params are valid");
    let argon2 = Argon2::new(Algorithm::Argon2id, Version::V0x13, params);
    let mut kek = [0u8; KEY_LEN];
    argon2
        .hash_password_into(password, salt, &mut kek)
        .map_err(|_| CryptoError::Aead)?;
    Ok(kek)
}

pub fn to_hex(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(bytes.len() * 2);
    for &b in bytes {
        out.push(HEX[(b >> 4) as usize] as char);
        out.push(HEX[(b & 0x0f) as usize] as char);
    }
    out
}

pub fn from_hex(s: &str) -> Result<Vec<u8>, CryptoError> {
    if !s.len().is_multiple_of(2) {
        return Err(CryptoError::Hex);
    }
    let mut out = Vec::with_capacity(s.len() / 2);
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let hi = hex_val(bytes[i])?;
        let lo = hex_val(bytes[i + 1])?;
        out.push((hi << 4) | lo);
        i += 2;
    }
    Ok(out)
}

fn hex_val(b: u8) -> Result<u8, CryptoError> {
    match b {
        b'0'..=b'9' => Ok(b - b'0'),
        b'a'..=b'f' => Ok(b - b'a' + 10),
        b'A'..=b'F' => Ok(b - b'A' + 10),
        _ => Err(CryptoError::Hex),
    }
}

pub fn check_name(name: &str) -> bool {
    if name.is_empty() || name.len() > 256 {
        return false;
    }
    if name.starts_with('/') || name.ends_with('/') || name.contains("..") {
        return false;
    }
    name.split('/').all(|seg| {
        !seg.is_empty()
            && seg.bytes().all(|b| {
                matches!(
                    b,
                    b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'.' | b'_' | b'@' | b'-'
                )
            })
    })
}

pub fn email_ok(raw: &str) -> Option<String> {
    let s = raw.trim().to_lowercase();
    if s.is_empty() || s.len() > 254 {
        return None;
    }
    let (local, domain) = s.split_once('@')?;
    if local.is_empty() || domain.is_empty() || domain.contains('@') || !domain.contains('.') {
        return None;
    }
    Some(s)
}

pub fn password_ok(password: &str) -> bool {
    let n = password.chars().count();
    (12..=256).contains(&n)
}

File: ui/src/lib/crypto.ts
/** Client AEAD and wraps. Same locked params as secd-core.
 *  The DEK lives only in a module binding for the tab's lifetime. */

import { xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { x25519 } from "@noble/curves/ed25519.js";
import { argon2id } from "@noble/hashes/argon2.js";

export const NONCE_LEN = 24;
export const KEY_LEN = 32;
export const SALT_LEN = 16;
export const ARGON2_M_KIB = 19_456;
export const ARGON2_T = 2;
export const ARGON2_P = 1;
export const ARGON2_VERSION = 0x13;
export const CONSOLE_TTL_MS = 12 * 60 * 60 * 1000;

const MIN_BLOB = NONCE_LEN + 1;
const ALG_X25519 = "x25519-xchacha20poly1305";

export type CryptoError = "key" | "aead" | "truncated" | "hex" | "rng" | "prf" | "factor";

export class CryptoFail extends Error {
  readonly code: CryptoError;
  constructor(code: CryptoError) {
    super(code);
    this.name = "CryptoFail";
    this.code = code;
  }
}

export type Factor = "passkey" | "password";

export type Wrap = {
  factor: Factor;
  blob: string;
  cred_id?: string;
  salt?: string;
};

let dek: Uint8Array | undefined;
let dekDeadline = 0;
let dekTimer: ReturnType<typeof setTimeout> | undefined;
const dekClearListeners = new Set<() => void>();

function fail(code: CryptoError): CryptoFail {
  return new CryptoFail(code);
}

export function zeroizeBytes(buf: Uint8Array): void {
  buf.fill(0);
}

export function mintDek(): Uint8Array {
  return randomBytes(KEY_LEN);
}

export function setDek(bytes: Uint8Array): void {
  if (bytes.length !== KEY_LEN) {
    throw fail("key");
  }
  dropDek();
  dek = new Uint8Array(KEY_LEN);
  dek.set(bytes);
  dekDeadline = Date.now() + CONSOLE_TTL_MS;
  dekTimer = setTimeout(() => {
    clearDek();
  }, CONSOLE_TTL_MS);
}

export function getDek(): Uint8Array | undefined {
  if (dek === undefined) {
    return undefined;
  }
  if (Date.now() >= dekDeadline) {
    clearDek();
    return undefined;
  }
  return dek;
}

/** Milliseconds until the tab drops its DEK; 0 when it holds none. */
export function dekRemainingMs(): number {
  return dek === undefined ? 0 : Math.max(0, dekDeadline - Date.now());
}

/** Fired after the tab DEK is dropped (TTL, sign-out, failed unwrap). Not on replace. */
export function onDekClear(fn: () => void): () => void {
  dekClearListeners.add(fn);
  return () => {
    dekClearListeners.delete(fn);
  };
}

function dropDek(): void {
  if (dekTimer !== undefined) {
    clearTimeout(dekTimer);
    dekTimer = undefined;
  }
  if (dek !== undefined) {
    zeroizeBytes(dek);
    dek = undefined;
  }
  dekDeadline = 0;
}

export function clearDek(): void {
  dropDek();
  for (const fn of [...dekClearListeners]) {
    fn();
  }
}

function randomBytes(n: number): Uint8Array {
  const out = new Uint8Array(n);
  if (globalThis.crypto === undefined || typeof globalThis.crypto.getRandomValues !== "function") {
    throw fail("rng");
  }
  globalThis.crypto.getRandomValues(out);
  return out;
}

function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

export function sealWithNonce(
  key: Uint8Array,
  name: string,
  plaintext: Uint8Array,
  nonce: Uint8Array,
): Uint8Array {
  if (key.length !== KEY_LEN || nonce.length !== NONCE_LEN) {
    throw fail("key");
  }
  const aad = utf8(name);
  const cipher = xchacha20poly1305(key, nonce, aad);
  const ct = cipher.encrypt(plaintext);
  const blob = new Uint8Array(NONCE_LEN + ct.length);
  blob.set(nonce, 0);
  blob.set(ct, NONCE_LEN);
  return blob;
}

export function seal(key: Uint8Array, name: string, plaintext: Uint8Array): Uint8Array {
  return sealWithNonce(key, name, plaintext, randomBytes(NONCE_LEN));
}

export function open(key: Uint8Array, name: string, blob: Uint8Array): Uint8Array {
  if (blob.length < MIN_BLOB) {
    throw fail("truncated");
  }
  if (key.length !== KEY_LEN) {
    throw fail("key");
  }
  const nonce = blob.subarray(0, NONCE_LEN);
  const ct = blob.subarray(NONCE_LEN);
  const aad = utf8(name);
  const cipher = xchacha20poly1305(key, nonce, aad);
  try {
    return cipher.decrypt(ct);
  } catch {
    throw fail("aead");
  }
}

function prfKek(prf: Uint8Array): Uint8Array {
  if (prf.length < KEY_LEN) {
    throw fail("prf");
  }
  return prf.subarray(0, KEY_LEN);
}

function derivePasswordKek(password: Uint8Array, salt: Uint8Array): Uint8Array {
  return argon2id(password, salt, {
    t: ARGON2_T,
    m: ARGON2_M_KIB,
    p: ARGON2_P,
    dkLen: KEY_LEN,
    version: ARGON2_VERSION,
  });
}

export function wrapPassword(dekBytes: Uint8Array, password: Uint8Array): Wrap {
  if (dekBytes.length !== KEY_LEN) {
    throw fail("key");
  }
  const salt = randomBytes(SALT_LEN);
  const kek = derivePasswordKek(password, salt);
  try {
    const blob = seal(kek, "password", dekBytes);
    return {
      factor: "password",
      blob: toHex(blob),
      salt: toHex(salt),
    };
  } finally {
    zeroizeBytes(kek);
    zeroizeBytes(salt);
  }
}

export function unwrapPassword(wrap: Wrap, password: Uint8Array): Uint8Array {
  if (wrap.factor !== "password") {
    throw fail("factor");
  }
  const saltHex = wrap.salt;
  if (saltHex === undefined) {
    throw fail("factor");
  }
  const salt = fromHex(saltHex);
  if (salt.length !== SALT_LEN) {
    zeroizeBytes(salt);
    throw fail("hex");
  }
  const kek = derivePasswordKek(password, salt);
  zeroizeBytes(salt);
  try {
    const blob = fromHex(wrap.blob);
    try {
      return open(kek, "password", blob);
    } finally {
      zeroizeBytes(blob);
    }
  } finally {
    zeroizeBytes(kek);
  }
}

export function wrapPasskey(dekBytes: Uint8Array, prf: Uint8Array, credId: string): Wrap {
  if (dekBytes.length !== KEY_LEN) {
    throw fail("key");
  }
  const kek = new Uint8Array(prfKek(prf));
  try {
    const blob = seal(kek, "passkey", dekBytes);
    return {
      factor: "passkey",
      blob: toHex(blob),
      cred_id: credId,
    };
  } finally {
    zeroizeBytes(kek);
  }
}

export function unwrapPasskey(wrap: Wrap, prf: Uint8Array): Uint8Array {
  if (wrap.factor !== "passkey") {
    throw fail("factor");
  }
  const kek = new Uint8Array(prfKek(prf));
  try {
    const blob = fromHex(wrap.blob);
    try {
      return open(kek, "passkey", blob);
    } finally {
      zeroizeBytes(blob);
    }
  } finally {
    zeroizeBytes(kek);
  }
}

export function wrapFromJson(v: unknown): Wrap | undefined {
  if (typeof v !== "object" || v === null) {
    return undefined;
  }
  const rec = v as Record<string, unknown>;
  const factor = rec["factor"];
  if (factor !== "passkey" && factor !== "password") {
    return undefined;
  }
  const blob = rec["blob"];
  if (typeof blob !== "string") {
    return undefined;
  }
  const out: Wrap = { factor, blob };
  if (typeof rec["cred_id"] === "string") {
    out.cred_id = rec["cred_id"];
  }
  if (typeof rec["salt"] === "string") {
    out.salt = rec["salt"];
  }
  return out;
}

export function wrapToJson(w: Wrap): Record<string, string> {
  const m: Record<string, string> = { factor: w.factor };
  if (w.cred_id !== undefined) {
    m["cred_id"] = w.cred_id;
  }
  if (w.salt !== undefined) {
    m["salt"] = w.salt;
  }
  m["blob"] = w.blob;
  return m;
}

export function wrapsFromJson(v: unknown): Wrap[] {
  if (typeof v !== "object" || v === null) {
    return [];
  }
  const arr = (v as { wraps?: unknown }).wraps;
  if (!Array.isArray(arr)) {
    return [];
  }
  const out: Wrap[] = [];
  for (const item of arr) {
    const w = wrapFromJson(item);
    if (w) {
      out.push(w);
    }
  }
  return out;
}

export function unwrapAny(
  wraps: Wrap[],
  password: Uint8Array | undefined,
  prf: Uint8Array | undefined,
): Uint8Array | undefined {
  if (prf !== undefined) {
    for (const w of wraps) {
      if (w.factor === "passkey") {
        try {
          return unwrapPasskey(w, prf);
        } catch {
          /* next wrap */
        }
      }
    }
  }
  if (password !== undefined) {
    for (const w of wraps) {
      if (w.factor === "password") {
        try {
          return unwrapPassword(w, password);
        } catch {
          /* next wrap */
        }
      }
    }
  }
  return undefined;
}

export function sealDekToEph(
  dekBytes: Uint8Array,
  theirPub: Uint8Array,
): { alg: string; eph_pub: string; blob: string } {
  if (dekBytes.length !== KEY_LEN || theirPub.length !== KEY_LEN) {
    throw fail("key");
  }
  const secret = randomBytes(KEY_LEN);
  const sk = new Uint8Array(secret);
  let shared: Uint8Array | undefined;
  try {
    const pub = x25519.getPublicKey(sk);
    const their = new Uint8Array(theirPub);
    shared = x25519.getSharedSecret(sk, their);
    zeroizeBytes(their);
    const blob = seal(shared, "dek", dekBytes);
    return {
      alg: ALG_X25519,
      eph_pub: toHex(pub),
      blob: toHex(blob),
    };
  } catch (e) {
    if (e instanceof CryptoFail) {
      throw e;
    }
    throw fail("key");
  } finally {
    zeroizeBytes(secret);
    zeroizeBytes(sk);
    if (shared !== undefined) {
      zeroizeBytes(shared);
    }
  }
}

export function x25519Shared(secret: Uint8Array, theirPub: Uint8Array): Uint8Array {
  if (secret.length !== KEY_LEN || theirPub.length !== KEY_LEN) {
    throw fail("key");
  }
  const sk = new Uint8Array(secret);
  const their = new Uint8Array(theirPub);
  try {
    return x25519.getSharedSecret(sk, their);
  } catch {
    throw fail("key");
  } finally {
    zeroizeBytes(sk);
    zeroizeBytes(their);
  }
}

export function x25519Public(secret: Uint8Array): Uint8Array {
  if (secret.length !== KEY_LEN) {
    throw fail("key");
  }
  const sk = new Uint8Array(secret);
  try {
    return x25519.getPublicKey(sk);
  } catch {
    throw fail("key");
  } finally {
    zeroizeBytes(sk);
  }
}

export function toHex(bytes: Uint8Array): string {
  const HEX = "0123456789abcdef";
  let out = "";
  for (const b of bytes) {
    out += HEX.charAt(b >> 4);
    out += HEX.charAt(b & 0x0f);
  }
  return out;
}

export function fromHex(s: string): Uint8Array {
  if (s.length % 2 !== 0) {
    throw fail("hex");
  }
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < s.length; i += 2) {
    const hi = hexVal(s.charCodeAt(i));
    const lo = hexVal(s.charCodeAt(i + 1));
    out[i / 2] = (hi << 4) | lo;
  }
  return out;
}

function hexVal(c: number): number {
  if (c >= 48 && c <= 57) {
    return c - 48;
  }
  if (c >= 97 && c <= 102) {
    return c - 97 + 10;
  }
  if (c >= 65 && c <= 70) {
    return c - 65 + 10;
  }
  throw fail("hex");
}

export function checkName(name: string): boolean {
  if (name.length === 0 || name.length > 256) {
    return false;
  }
  if (name.startsWith("/") || name.endsWith("/") || name.includes("..")) {
    return false;
  }
  return name.split("/").every((seg) => {
    if (seg.length === 0) {
      return false;
    }
    for (let i = 0; i < seg.length; i++) {
      const b = seg.charCodeAt(i);
      const ok =
        (b >= 65 && b <= 90) ||
        (b >= 97 && b <= 122) ||
        (b >= 48 && b <= 57) ||
        b === 46 ||
        b === 95 ||
        b === 64 ||
        b === 45;
      if (!ok) {
        return false;
      }
    }
    return true;
  });
}

export function emailOk(raw: string): string | undefined {
  const s = raw.trim().toLowerCase();
  if (s.length === 0 || s.length > 254) {
    return undefined;
  }
  const at = s.indexOf("@");
  if (at < 0) {
    return undefined;
  }
  const local = s.slice(0, at);
  const domain = s.slice(at + 1);
  if (local.length === 0 || domain.length === 0 || domain.includes("@") || !domain.includes(".")) {
    return undefined;
  }
  return s;
}

export function passwordOk(password: string): boolean {
  const n = [...password].length;
  return n >= 12 && n <= 256;
}
